common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala

```diff
@@ -0,0 +1,106 @@
+  MemoryMetricsGetter
+}
+import org.apache.http.client.HttpClient
+import org.apache.http.impl.client.DefaultHttpClient
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+
+import java.util.concurrent.atomic.AtomicBoolean
+import scala.collection.{immutable, mutable}
+import scala.concurrent.duration.{FiniteDuration, SECONDS}
+
+class MetricsCollector(hostname: String, sparkConf: SparkConf) extends Logging {
+  private val dmAppId = DmAppId(Configs.getDMAppId(sparkConf))
+  private val collectorURL = Configs.collectorUrl(sparkConf).stripSuffix("/")
+  private val accessTokenOption = Configs.accessTokenOption(sparkConf)
+
+  private val started: AtomicBoolean = new AtomicBoolean(false)
+  private val MetricsQueue = new mutable.Queue[MetricsEvent]()
+  private val httpClientMetrics: HttpClient = new DefaultHttpClient()
+  private val threads = new mutable.ListBuffer[Thread]()
+
+  private def publishMetrics(payload: MetricsPayload): Unit = {
+    val url = s"$collectorURL/metrics"
+
+    try {
+      sendRequest(
+        httpClientMetrics,
+        url,
+        accessTokenOption.get,
+        payload.toJson,
+        "Metrics payload sent successfully"
+      )
+    } catch {
+      case e: Exception =>
+        logWarning(s"Failed to send metrics to $url: ${e.getMessage}")
+        throw e
+    }
+  }
+
+  private def sendMetrics(): Unit = {
+    val memoryMetrics = MetricsQueue
+      .synchronized(MetricsQueue.take(1000))
+      .to[immutable.Seq]
+      .map(_.toString)
+    val payloadSize = memoryMetrics.length
+    if (payloadSize > 0) {
+      try {
+        publishMetrics(
+          MetricsPayload(
+            dmAppId,
+            memoryMetrics
+          )
+        )
+        MetricsQueue.synchronized {
+          for (_ <- 1 to payloadSize) MetricsQueue.dequeue()
+        }
+      } catch {
+        case _: Exception =>
+          Thread.sleep(5000)
+      }
+    }
+  }
+
+  def startIfNecessary(): Unit = {
+    if (started.compareAndSet(false, true)) {
+      threads += startRepeatThread(FiniteDuration(1, SECONDS)) {
+        logDebug("Logged memory metrics Poller")
+        val metrics = MemoryMetricsGetter.get().computeAllMetrics()
+        val event = MemoryMetricsEvent(hostname, metrics)
+        MetricsQueue.synchronized {
+          MetricsQueue.enqueue(event)
+        }
+      }
+      logInfo("Started MemoryMetrics Poller thread")
+      threads += startRepeatThread(FiniteDuration(2, SECONDS)) {
+        logDebug("Logged metrics Sender")
+        sendMetrics()
+      }
+      logInfo("Started MemoryMetrics Sender thread")
+    }
+  }
+
+  def stop(): Unit = {
+    threads.foreach(_.stop())
+  }
+}
+
+object MetricsCollector {
+
+  private var sharedCollector: Option[MetricsCollector] = None
+
+  def getOrCreate(hostname: String, sparkConf: SparkConf): MetricsCollector = {
+    if (sharedCollector.isEmpty) {
+      sharedCollector = Option(new MetricsCollector(hostname, sparkConf))
```

Comment:
Wonder if making `sharedCollector` a `lazy val` wouldn't be the same yet less code?

```
lazy val sharedCollector = new MetricsCollector(hostname, sparkConf)
```

Reply:
How do you pass the params to the object  if you want to use a lazy val ?

Reply:
Correct! Can you change `Option` to `Some`?